README
NAME
    Perinci::To::POD - Generate POD documentation for a package from Rinci
    metadata

VERSION
    version 0.32

SYNOPSIS
    You can use the included peri-pkg-doc script, or:

     use Perinci::To::POD;
     my $doc = Perinci::To::POD->new(url => "/Some/Module/");
     say $doc->gen_doc;

    To generate documentation for a single function, see
    Perinci::Sub::To::POD or the provided command-line script peri-func-doc.

    To generate a usage-like help message for a single function, you can try
    the peri-func-usage from the Perinci::CmdLine distribution.

AUTHOR
    Steven Haryanto <stevenharyanto@gmail.com>

COPYRIGHT AND LICENSE
    This software is copyright (c) 2013 by Steven Haryanto.

    This is free software; you can redistribute it and/or modify it under
    the same terms as the Perl 5 programming language system itself.

DESCRIPTION
FUNCTIONS
    None are exported by default, but they are exportable.



Changes
Revision history for Perinci-To-POD

0.31	 2013-07-26 (SHARYANTO)

         Update function name.


0.30	 2013-07-26 (SHARYANTO)

         - Split bin/peri-doc to bin/peri-pkg-doc and bin/peri-func-doc (in
           Perinci-Sub-To-POD).

         - bin/peri-pkg-doc: Allow enabling logging using LOG=1.

         - Update to using Perinci::To::PackageBase 0.20.


0.29	 2013-07-17 (SHARYANTO)

	 [BUG FIXES]

         - bin/peri-doc: Make --format work (previously it's overridden by
           Perinci::CmdLine's common option --format).


0.28	 2012-09-05 (SHARYANTO)

	 Fix list of special arguments for transaction.


0.27	 2012-09-05 (SHARYANTO)

	 Update to Rinci 1.1.29 (transaction protocol v=2, older v not
	 supported).


0.26	 2012-06-22 (SHARYANTO)

	 Update to Rinci 1.1.11 (pkg_version -> entity_version).


0.25	 2012-06-20 (SHARYANTO)

	 Explain possible special arguments.


0.24	 2012-06-20 (SHARYANTO)

	 Explain function features.


0.23	 2012-03-15 (SHARYANTO)

	 No functional change. Update to using newer Perinci::To::PackageBase.


0.22	 2012-03-14 (SHARYANTO)

	 No functional change. Update to using Perinci::To::PackageBase.


0.21	 2012-03-13 (SHARYANTO)

	 First release, renamed from Sub-Spec-To-Pod. Now follow Rinci/Riap
	 specification.


lib/Perinci/To/POD.pm
package Perinci::To::POD;

use 5.010001;
use Log::Any '$log';
use Moo;

extends 'Perinci::To::PackageBase';

our $VERSION = '0.32'; # VERSION

sub BUILD {
    my ($self, $args) = @_;
}

sub _md2pod {
    require Markdown::Pod;

    my ($self, $md) = @_;
    state $m2p = Markdown::Pod->new;
    $m2p->markdown_to_pod(markdown => $md);
}

sub gen_doc_section_summary {
    my ($self) = @_;

    $self->SUPER::gen_doc_section_summary;
    my $res = $self->{_res};

    my $name_summary = join(
        "",
        $res->{name} // "",
        ($res->{name} && $res->{summary} ? ' - ' : ''),
        $res->{summary} // ""
    );

    $self->add_doc_lines(
        "=head1 " . uc($self->loc("Name")),
        "",
        $name_summary,
        "",
    );
}

sub gen_doc_section_version {
    my ($self) = @_;

    $self->add_doc_lines(
        "=head1 " . uc($self->loc("Version")),
        "",
        $self->{_meta}{entity_version} // '?',
        "",
    );
}

sub gen_doc_section_description {
    my ($self) = @_;

    $self->add_doc_lines(
        "=head1 " . uc($self->loc("Description")),
        ""
    );

    $self->SUPER::gen_doc_section_description;
    my $res = $self->{_res};

    if ($res->{description}) {
        $self->add_doc_lines(
            $self->_md2pod($res->{description}),
            "",
        );
    }

    #$self->add_doc_lines(
    #    $self->loc("This module has L<Rinci> metadata") . ".",
    #    "",
    #);
}

sub gen_doc_section_functions {
    require Perinci::Sub::To::POD;

    my ($self) = @_;
    my $res = $self->{_res};

    $self->{_fgen} //= Perinci::Sub::To::POD->new(
        _pa => $self->_pa, # to avoid multiple instances of pa objects
    );

    $self->add_doc_lines(
        "=head1 " . uc($self->loc("Functions")),
        "",
    );

    $self->SUPER::gen_doc_section_functions;

    # temporary, since we don't parse export information yet
    $self->add_doc_lines(
        $self->loc("None are exported by default, but they are exportable."),
        "",
    );

    # XXX if module uses Perinci::Exporter, show a basic usage for importing

    # XXX categorize functions based on tags
    for my $furi (sort keys %{ $res->{functions} }) {
        my $fname;
        for ($fname) { $_ = $furi; s!.+/!! }
        for (@{ $res->{functions}{$furi} }) {
            chomp;
            $self->add_doc_lines($_);
        }
    }
}

1;
# ABSTRACT: Generate POD documentation for a package from Rinci metadata

__END__

=pod

=encoding utf-8

=head1 NAME

Perinci::To::POD - Generate POD documentation for a package from Rinci metadata

=head1 VERSION

version 0.32

=head1 SYNOPSIS

You can use the included L<peri-pkg-doc> script, or:

 use Perinci::To::POD;
 my $doc = Perinci::To::POD->new(url => "/Some/Module/");
 say $doc->gen_doc;

To generate documentation for a single function, see L<Perinci::Sub::To::POD>
or the provided command-line script L<peri-func-doc>.

To generate a usage-like help message for a single function, you can try
the L<peri-func-usage> from the L<Perinci::CmdLine> distribution.

=for Pod::Coverage .+

=head1 AUTHOR

Steven Haryanto <stevenharyanto@gmail.com>

=head1 COPYRIGHT AND LICENSE

This software is copyright (c) 2013 by Steven Haryanto.

This is free software; you can redistribute it and/or modify it under
the same terms as the Perl 5 programming language system itself.

=head1 DESCRIPTION

=head1 FUNCTIONS


None are exported by default, but they are exportable.

=cut


lib/Perinci/To/POD/I18N/en.pm
package Perinci::To::POD::I18N::en;
use parent qw(Perinci::To::POD::I18N Perinci::To::PackageBase::I18N::en);

use Locale::Maketext::Lexicon::Gettext;
our %Lexicon = %{ Locale::Maketext::Lexicon::Gettext->parse(<DATA>) };

our $VERSION = '0.32'; # VERSION

#use Data::Dump; dd \%Lexicon;

1;
# ABSTRACT: English translation for Perinci::To::POD

=pod

=encoding utf-8

=head1 NAME

Perinci::To::POD::I18N::en - English translation for Perinci::To::POD

=head1 VERSION

version 0.32

=head1 AUTHOR

Steven Haryanto <stevenharyanto@gmail.com>

=head1 COPYRIGHT AND LICENSE

This software is copyright (c) 2013 by Steven Haryanto.

This is free software; you can redistribute it and/or modify it under
the same terms as the Perl 5 programming language system itself.

=head1 DESCRIPTION

=head1 FUNCTIONS


None are exported by default, but they are exportable.

=cut

__DATA__

msgid  "This module has L<Rinci> metadata"
msgstr "This module has L<Rinci> metadata"

msgid  "No arguments"
msgstr "No arguments"

# function

msgid  "None are exported by default, but they are exportable."
msgstr "None are exported by default, but they are exportable."

# function's special arguments

msgid  "Pass -reverse=>1 to reverse operation."
msgstr "Pass -reverse=>1 to reverse operation."

msgid  "To undo, pass -undo_action=>'undo' to function. You will also need to pass -undo_data. For more details on undo protocol, see L<Rinci::Undo>."
msgstr "To undo, pass -undo_action=>'undo' to function. You will also need to pass -undo_data. For more details on undo protocol, see L<Rinci::Undo>."

msgid  "Required if you pass -undo_action=>'undo'. For more details on undo protocol, see L<Rinci::Undo>."
msgstr "Required if you pass -undo_action=>'undo'. For more details on undo protocol, see L<Rinci::Undo>."

msgid  "Pass -dry_run=>1 to enable simulation mode."
msgstr "Pass -dry_run=>1 to enable simulation mode."

msgid  "This function supports transactions."
msgstr "This function supports transactions."

msgid  "For more information on transaction, see L<Rinci::Transaction>."
msgstr "For more information on transaction, see L<Rinci::Transaction>."


Build.PL

use strict;
use warnings;

use Module::Build 0.3601;


my %module_build_args = (
  "build_requires" => {
    "Module::Build" => "0.3601"
  },
  "configure_requires" => {
    "Module::Build" => "0.3601"
  },
  "dist_abstract" => "Generate POD documentation for a package from Rinci metadata",
  "dist_author" => [
    "Steven Haryanto <stevenharyanto\@gmail.com>"
  ],
  "dist_name" => "Perinci-To-POD",
  "dist_version" => "0.32",
  "license" => "perl",
  "module_name" => "Perinci::To::POD",
  "recommends" => {},
  "recursive_test_files" => 1,
  "requires" => {
    "File::Which" => 0,
    "Locale::Maketext::Lexicon::Gettext" => 0,
    "Log::Any" => 0,
    "Markdown::Pod" => 0,
    "Moo" => 0,
    "Perinci::CmdLine" => "0.80",
    "Perinci::Sub::To::POD" => 0,
    "Perinci::To::PackageBase" => "0.20",
    "Perinci::To::PackageBase::I18N" => "0.20",
    "Perinci::To::PackageBase::I18N::en" => "0.20",
    "Perinci::To::Text" => "0.20",
    "Riap" => "v1.1.8",
    "Rinci" => "v1.1.29",
    "perl" => "5.010001"
  },
  "script_files" => [
    "bin/peri-pkg-doc",
    "bin/peri-doc"
  ],
  "test_requires" => {}
);


unless ( eval { Module::Build->VERSION(0.4004) } ) {
  my $tr = delete $module_build_args{test_requires};
  my $br = $module_build_args{build_requires};
  for my $mod ( keys %$tr ) {
    if ( exists $br->{$mod} ) {
      $br->{$mod} = $tr->{$mod} if $tr->{$mod} > $br->{$mod};
    }
    else {
      $br->{$mod} = $tr->{$mod};
    }
  }
}

my $build = Module::Build->new(%module_build_args);

$build->create_build_script;
